src/ligero/mod.rs

```diff
@@ -14,26 +13,114 @@ pub mod verifier;
-    pub row_size: usize,
-    /// The total size of a tableau row.
-    /// There is some confusion in the specification between this and BLOCK.
-    pub num_columns: u128,
+    /// The size of a block, in terms of number of field elements. Also `BLOCK`. The specification
+    /// describes this quantity as the "size of each row", but that would be `NCOL` or
+    /// `num_columns`.
+    pub block_size: usize,
+    /// The total size of a tableau row. Also `NCOL`.
+    pub num_columns: usize,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TableauLayout<'a> {
+    parameters: &'a LigeroParameters,
+    num_witnesses: usize,
+    num_quadratic_constraints: usize,
 }
 
-impl LigeroParameters {
-    /// DBLOCK, 2 * BLOCK - 1
+impl<'a> TableauLayout<'a> {
+    pub fn new(
+        parameters: &'a LigeroParameters,
+        num_witnesses: usize,
+        num_quadratic_constraints: usize,
+    ) -> Self {
+        Self {
+            parameters,
+            num_witnesses,
+            num_quadratic_constraints,
+        }
+    }
+
+    /// The size of a block, in terms of number of field elements. Also `BLOCK`. The specification
+    /// describes this quantity as the "size of each row", but that would be `NCOL` or
+    /// `num_columns`.
+    pub fn block_size(&self) -> usize {
+        self.parameters.block_size
+    }
+
+    /// The total size of a tableau row. Also `NCOL`.
+    pub fn num_columns(&self) -> usize {
+        self.parameters.num_columns
+    }
+
+    /// The number of columns of the commitment matrix that the Verifier requests to be revealed by
+    /// the Prover. Also `NREQ`.
+    pub fn nreq(&self) -> usize {
+        self.parameters.nreq
+    }
+
+    /// `DBLOCK = 2 * BLOCK - 1`
     pub fn dblock(&self) -> usize {
-        self.row_size * 2 - 1
+        self.parameters.block_size * 2 - 1
+    }
+
+    /// The number of witness values included in each row. Also `WR`.
+    pub fn witnesses_per_row(&self) -> usize {
+        self.parameters.witnesses_per_row
+    }
+
+    /// The number of quadratic constraints written in each row. Also `QR`.
+    pub fn quadratic_constraints_per_row(&self) -> usize {
+        self.parameters.quadratic_constraints_per_row
+    }
+
+    /// Index of the first row of the tableau containing linear constraints on the witnesses.
+    pub fn first_linear_constraint_row(&self) -> usize {
+        // One row each for low degree, linear and quadratic tests.
+        3
+    }
+
+    /// Index of the first row of the tableau containing quadratic constraints on the witnesses.
+    pub fn first_quadratic_constraint_row(&self) -> usize {
+        self.first_linear_constraint_row() + self.num_linear_constraint_rows()
+    }
+
+    /// The number of triples of tableau rows needed to represent the quadratic constraints
+    pub fn num_quadratic_triples(&self) -> usize {
+        self.num_quadratic_constraints
+            .div_ceil(self.parameters.witnesses_per_row)
```

Comment:
We should be using this parameter instead.
```suggestion
            .div_ceil(self.parameters.quadratic_constraints_per_row)
```

Reply:
https://github.com/google/longfellow-zk/issues/106